### PR TITLE
Disconnected OCS cluster installation from internal builds

### PR DESCRIPTION
**Pull request by dahorak**
This is initial implementation of OCS installation in disconnected environment from internal builds.

Related ticket https://projects.engineering.redhat.com/browse/OCSQE-189
This also requires following changes: [flexy-ocs-private MR!10](https://gitlab.cee.redhat.com/ocs/flexy-ocs-private/-/merge_requests/10)

Signed-off-by: Daniel Horak <dahorak@redhat.com>

### CODE REVIEW

**Comment by dahorak** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: deployment
> PR Test Path: tests/
> Additional Test Params:
> OCP VERSION: 4.6
> OCS VERSION: 4.6
> 
> [Job PASSED](https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-trigger-test-pr/3660//testReport).

This is validation of normal installation (not disconnected).

**Comment by dahorak** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration: conf/ocsci/aws_upi_flexy_rhcos_3m_3w_disconnected.yaml
> PR Test Suite: tier1
> PR Test Path: tests/
> Additional Test Params:
> OCP VERSION: 4.6
> OCS VERSION: 4.6
> 
> [Job FAILED](https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-trigger-test-pr/3716//testReport) (installation failed, tests not executed).

This cluster was unexpectedly deleted by AWS cleanup job in the middle of test execution.

**Comment by dahorak** [comment]
Verification tier1 tests run against disconnected cluster: https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/16369/testReport/
Lot's of mcg tests are failing, will debug them and prospectively send another update here or in new PR.

File: ocs_ci/helpers/helpers.py [hunk 1]
```diff
@@ -115,6 +115,36 @@ def wait_for_resource_state(resource, state, timeout=60):
     logger.info(f"{resource.kind} {resource.name} reached state {state}")
 
 
+def update_container_with_proxy_env(job_pod_dict):
```
Comment (petr-balogh) [inline ocs_ci/helpers/helpers.py:118]: Maybe move to some proxy-helpers module as it's not that general function to be in just helpers.py

**Review by petr-balogh — DISMISSED** [review]

File: ocs_ci/helpers/helpers.py (hunk 1 above)
Comment (dahorak) [reply ocs_ci/helpers/helpers.py]: Moved to new module `ocs_ci.helpers.proxy` in 8acb403.

File: ocs_ci/deployment/flexy.py [hunk 2]
```diff
@@ -325,21 +329,51 @@ def flexy_post_processing(self):
...
+            )
+            logger.info(ssh_int_svc.exec_cmd(cmd=cmd))
+            cmd = (
+                """sudo sed -i '1i acl ocs_whitelist dstdomain """
+                f"""{" ".join(constants.DISCON_CL_PROXY_ALLOWED_DOMAINS)}' """
+                """/srv/squid/etc/squid.conf"""
+            )
+            logger.info(ssh_int_svc.exec_cmd(cmd=cmd))
+            cmd = "sudo systemctl restart squid-proxy.service"
+            logger.info(ssh_int_svc.exec_cmd(cmd=cmd))
```
[the remark this answers is not in the document]
Comment (dahorak) [reply ocs_ci/deployment/flexy.py]: Moved to existing module `ocs_ci.utility.flexy` (I think it fits there as it is tightly connected to the Flexy installation): 8acb403.

**Comment by dahorak** [comment]
Rebased and resolved conflicts.

**Review by petr-balogh — DISMISSED** [review]

**Review by petr-balogh — APPROVED** [review]

**Comment by dahorak** [comment]
[This is last verification job](https://ocs4-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/qe-deploy-ocs-cluster/16626/testReport/) - OCP 4.7, OCS 4.7, tier1 test suite (failures from `mcg` tests are _expected_).

**Review by clacroix12 — APPROVED** [review]
Looks great to me! Only a few very minor docstring issues that shouldn't hold up the merge.

File: ocs_ci/deployment/disconnected.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+
+def prepare_disconnected_ocs_deployment():
+    """
+    Prepare disconnected ocs deployment:
+    - get related images from OCS operator bundle csv
+    - mirror related images to mirror registry
+    - create imageContentSourcePolicy for the mirrored images
+    - disable the default OperatorSources
+    """
```
[the remark this answers is not in the document]
Comment (dahorak) [reply ocs_ci/deployment/disconnected.py:63]: I've already spotted this one and have it fixed in my code, I'll add it to the following PR.

File: ocs_ci/deployment/flexy.py [hunk 4]
```diff
@@ -297,49 +302,66 @@ def flexy_prepare_work_dir(self):
...
+        # mirror install-dir to cluster path (auth directory, metadata.json
+        # file and other files)
+        install_dir = os.path.join(self.flexy_host_dir, "flexy/workdir/install-dir/")
+        rsync_cmd = f"rsync -av {install_dir} {self.cluster_path}/"
+        exec_cmd(rsync_cmd)
 
     def flexy_post_processing(self):
         """
-        Update global pull-secret and configure ntp (if required).
+        Pefrom few actions required after flexy execution:
```
[the remark this answers is not in the document]
Comment (dahorak) [reply ocs_ci/deployment/flexy.py:328]: I'll fix this in following PR.

File: ocs_ci/utility/utils.py [hunk 5]
```diff
@@ -2359,70 +2360,151 @@ def convert_device_size(unformatted_size, units_to_covert_to):
...
     """
-    Mirror image to mirror image registry.
+    Prepare customized pull-secret containing auth section related to given
+    image(s). If image(s) not defined or no related section is found, it will
+    use whole content of pull-secret.
 
     Args:
-        image (str): image to be mirrored, can be defined just with name or
-            with full url, with or without tag or digest
+        images (str or list): image (or images) to match with auth section
```
[the remark this answers is not in the document]
Comment (dahorak) [reply ocs_ci/utility/utils.py:2370]: I'll fix this in following PR.

**Pull request merged**